runtime/src/snapshot_utils.rs

```diff
@@ -1769,6 +1803,102 @@ fn rebuild_bank_from_snapshots(
+    verify_slot_deltas_with_history(slot_deltas, &bank.get_slot_history(), bank.slot())
+}
+
+/// Verify that the snapshot's slot deltas are not corrupt/invalid
+/// These checks are simple/structural
+fn verify_slot_deltas_structural(
+    slot_deltas: &[BankSlotDelta],
+    bank_slot: Slot,
+) -> std::result::Result<(), VerifySlotDeltasError> {
+    // there should not be more entries than that status cache's max
+    let num_entries = slot_deltas.len();
+    if num_entries > status_cache::MAX_CACHE_ENTRIES {
+        return Err(VerifySlotDeltasError::TooManyEntries(
+            num_entries,
+            status_cache::MAX_CACHE_ENTRIES,
+        ));
+    }
+
+    let mut slots_seen_so_far = HashSet::new();
+    for &(slot, is_root, ..) in slot_deltas {
+        // all entries should be roots
+        if !is_root {
+            return Err(VerifySlotDeltasError::SlotIsNotRoot(slot));
+        }
+
+        // all entries should be for slots less than or equal to the bank's slot
+        if slot > bank_slot {
+            return Err(VerifySlotDeltasError::SlotGreaterThanMaxRoot(
+                slot, bank_slot,
+            ));
+        }
+
+        // there should only be one entry per slot
+        let inserted = slots_seen_so_far.insert(slot);
+        if !inserted {
+            return Err(VerifySlotDeltasError::SlotHasMultipleEntries(slot));
+        }
+    }
+
+    Ok(())
+}
+
+/// Verify that the snapshot's slot deltas are not corrupt/invalid
+/// These checks use the slot history for verification
+fn verify_slot_deltas_with_history(
+    slot_deltas: &[BankSlotDelta],
+    slot_history: &SlotHistory,
+    bank_slot: Slot,
+) -> std::result::Result<(), VerifySlotDeltasError> {
+    // all entries should be in the bank's slot history
+    let entry_missing_from_history = slot_deltas
+        .iter()
+        .find(|(slot, ..)| slot_history.check(*slot) != Check::Found);
+    if let Some((slot, ..)) = entry_missing_from_history {
+        return Err(VerifySlotDeltasError::SlotNotFoundInHistory(*slot));
+    }
+
+    // all slots in the history should be in the slot deltas (up to MAX_CACHE_ENTRIES)
+    // this ensures nothing was removed from the status cache
+
+    fn slot_deltas_contains_slot(slot_deltas: &[BankSlotDelta], slot: Slot) -> bool {
+        slot_deltas
+            .iter()
+            .any(|(slot_from_deltas, ..)| *slot_from_deltas == slot)
+    }
+
+    // go through the slot history and make sure there's an entry for each slot
+    // note: it's important to go highest-to-lowest since the status cache is a fifo
+    let mut entries_checked = 0;
+    for slot in (0..=bank_slot).rev() {
+        if entries_checked >= status_cache::MAX_CACHE_ENTRIES {
+            // We've gone through the whole slot deltas; done!
+            break;
+        }
+
+        if slot_history.check(slot) != Check::Found {
+            // This slot is not in the slot history; no need to check slot deltas
+            continue;
+        }
```

Comment:
nit: i don't know you like this iter kanfoo to replace this manual loop manipulation.

```rust
// go through the slot history and make sure there's an entry for each slot, upto MAX_CACHE_ENTRIES, skipping non-existing ones in the slot history.
// note: it's important to go highest-to-lowest since the status cache is a fifo
for slot_in_history in (0..=bank_slot).rev().filter(|s| slot_history.check(slot) == Check::Found).take(status_cache::MAX_CACHE_ENTRIES)
    if !slot_deltas_contains_slot(slot_deltas, slot) {
        ...
    }
}
```

also, i think you can further go with `::collect<Result<_>>` if you like the style. :)